espnet/nets/pytorch_backend/transformer/encoder_layer.py

```diff
@@ -36,34 +42,62 @@ def __init__(
         size,
         self_attn,
         feed_forward,
+        conv_module,
         dropout_rate,
         normalize_before=True,
         concat_after=False,
+        macaron_style=False,
     ):
         """Construct an EncoderLayer object."""
         super(EncoderLayer, self).__init__()
         self.self_attn = self_attn
         self.feed_forward = feed_forward
-        self.norm1 = LayerNorm(size)
-        self.norm2 = LayerNorm(size)
+        self.ff_scale = 1.0
+        self.conv_module = conv_module
+        self.macaron_style = macaron_style
+        self.norm_ff = LayerNorm(size)  # for the FNN module
+        self.norm_mha = LayerNorm(size)  # for the MHA module
+        if self.macaron_style:
+            self.feed_forward_macaron = copy.deepcopy(feed_forward)
```

Comment:
why did you use deep copy instead of giving a pointer of `feed_forward` (this means sharing the parameters and it would be wrong though) or newly introduce the liner layer?

Reply:
The previous implementation of Conformer construct encoder likes this:
```
self.encoders = repeat(
    num_blocks,
    lambda: EncoderLayer(
        attention_dim,
        attention_layer(attention_heads, attention_dim, attention_dropout_rate),
        positionwise_layer(*positionwise_layer_args),
        positionwise_layer(*positionwise_layer_args),
        convolution_layer(*convolution_layer_args) if use_cnn_module else None,
        dropout_rate,
        normalize_before,
        concat_after,
        rel_pos,
        macaron_style,
    ),
)
```
However, Transformer doesn't need another feed-forward layer. So I have to define another layer inside the EncoderLayer(). I think deepcopy will create a new object.

Reply:
I see, but why don’t you use `torch.nn.Linear` to create an object?
You can obtain the dimension information from `feed_forward`.
I think deepcopy would be OK, but this irregular init method is not consistent with the other ways, and we might introduce some bugs.

Reply:
We didn't pass these arguments `attention_dim, linear_units, dropout_rate` to `EncoderLayer()`. Maybe I can use the same way as the convolution layer, like this
```
self.encoders = repeat(
    num_blocks,
    lambda: EncoderLayer(
        attention_dim,
        encoder_attn_layer(*encoder_attn_layer_args),
        positionwise_layer(*positionwise_layer_args),
        positionwise_layer(*positionwise_layer_args) if use_cnn_module else None,  # for macaron style
        convolution_layer(*convolution_layer_args) if use_cnn_module else None,
        dropout_rate,
        normalize_before,
        concat_after,
        macaron_style,
    ),
)
```

Reply:
Besides, the `positionwise_layer` contains 2 linear layers and a dropout layer. Construct a new `positionwise_layer ` will need some redundant codes.